src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoaderCursor.cs

```diff
@@ -0,0 +1,276 @@
+                }
+
+                return getterDelegate as ValueGetter<TValue>;
+            }
+
+            private ValueGetter<bool> CreateBooleanGetterDelegate(ColInfo colInfo)
+            {
+                int columnIndex = GetColumnIndex(colInfo);
+                return (ref bool value) => value = _input.GetBoolean(columnIndex);
+            }
+
+            private ValueGetter<byte> CreateByteGetterDelegate(ColInfo colInfo)
+            {
+                int columnIndex = GetColumnIndex(colInfo);
+                return (ref byte value) => value = _input.GetByte(columnIndex);
+            }
+
+            private ValueGetter<DateTime> CreateDateTimeGetterDelegate(ColInfo colInfo)
+            {
+                int columnIndex = GetColumnIndex(colInfo);
+                return (ref DateTime value) => value = _input.GetDateTime(columnIndex);
+            }
+
+            private ValueGetter<double> CreateDoubleGetterDelegate(ColInfo colInfo)
+            {
+                int columnIndex = GetColumnIndex(colInfo);
+                return (ref double value) => value = _input.GetDouble(columnIndex);
+            }
+
+            private ValueGetter<short> CreateInt16GetterDelegate(ColInfo colInfo)
+            {
+                int columnIndex = GetColumnIndex(colInfo);
+                return (ref short value) => value = _input.GetInt16(columnIndex);
+            }
+
+            private ValueGetter<int> CreateInt32GetterDelegate(ColInfo colInfo)
+            {
+                int columnIndex = GetColumnIndex(colInfo);
+                return (ref int value) => value = _input.GetInt32(columnIndex);
+            }
+
+            private ValueGetter<long> CreateInt64GetterDelegate(ColInfo colInfo)
+            {
+                int columnIndex = GetColumnIndex(colInfo);
+                return (ref long value) => value = _input.GetInt64(columnIndex);
+            }
+
+            private ValueGetter<sbyte> CreateSByteGetterDelegate(ColInfo colInfo)
+            {
+                int columnIndex = GetColumnIndex(colInfo);
+                return (ref sbyte value) => value = (sbyte)_input.GetByte(columnIndex);
+            }
+
+            private ValueGetter<float> CreateSingleGetterDelegate(ColInfo colInfo)
+            {
+                int columnIndex = GetColumnIndex(colInfo);
+                return (ref float value) => value = _input.GetFloat(columnIndex);
+            }
+
+            private ValueGetter<ReadOnlyMemory<char>> CreateStringGetterDelegate(ColInfo colInfo)
+            {
+                int columnIndex = GetColumnIndex(colInfo);
+                return (ref ReadOnlyMemory<char> value) => value = _input.GetString(columnIndex).AsMemory();
+            }
+
+            private ValueGetter<ushort> CreateUInt16GetterDelegate(ColInfo colInfo)
+            {
+                int columnIndex = GetColumnIndex(colInfo);
+                return (ref ushort value) => value = (ushort)_input.GetInt16(columnIndex);
```

Comment:
Are there no such things as unsigned integers in DbDataReader?

Reply:
Not that I saw. I would guess `DbDataReader` was made back when people cared about `CLSCompliant`, which the unsigned types aren't.

Reply:
Hmm, it is odd that `System.Data.DbType` has unsigned integers, but `DbDataReader` doesn't seem to support them. Oh well, I guess this is the best we can do.

---
In reply to: [308375777](https://github.com/dotnet/machinelearning/pull/4035#discussion_r308375777) [](ancestors = 308375777)